tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/Session.java

```diff
@@ -485,6 +484,51 @@ public void run(Op op) {
      * this field may be replaced by more type-safe equivalents at any time.
      */
     public RunMetadata metadata;
+
+    /**
+     * Current `pop` index in the `outputs` list.
+     */
+    private int index = 0;
+
+    Run(List<Tensor<?>> outputs, RunMetadata metadata) {
+      this.outputs = outputs;
+      this.metadata = metadata;
+    }
+
+    Run(List<Tensor<?>> outputs) {
+      this.outputs = outputs;
+    }
+
+    public <T extends TType> Tensor<T> pop(DataType<T> dtype) {
+      return outputs.get(index++).expect(dtype);
+    }
+
+    public int popInt(long... coordinates) {
+      return pop(TInt32.DTYPE).data().getInt(coordinates);
+    }
+
+    public long popLong(long... coordinates) {
+      return pop(TInt64.DTYPE).data().getLong(coordinates);
+    }
+
+    public float popFloat16(long... coordinates) {
+      return pop(TFloat16.DTYPE).data().getFloat(coordinates);
+    }
+
+    public float popFloat(long... coordinates) {
+      return pop(TFloat32.DTYPE).data().getFloat(coordinates);
+    }
+
+    public double popDouble(long... coordinates) {
+      return pop(TFloat64.DTYPE).data().getDouble(coordinates);
+    }
+
+    @Override
+    public void close() {
+      for (Tensor<?> tensor : this.outputs) {
+        tensor.close();
+      }
+    }
```

Comment:
Mmh, I was expecting those changes to come in, from [our conversation](https://github.com/tensorflow/java/pull/30#discussion_r408318991) with @Craigacp , I thought we were more heading on just finding a way to prevent to user to catch an exception as part of the control flow. 

I don't discard this idea of "popping" the tensors out of the `Run` object but there might be more thinking to do around this and it is not required for supporting the Data API.

Reply:
Ah, I was playing around with this, since it sounds like we will require some extra machinery, for graph mode datasets, but I'll move it to a separate branch.

Reply:
I removed the `pop()` methods from `Session.Run` for this PR; but later on it would be great to find a way to simplify the data-extraction-chain.

`Run.outputs.get(...).expect(...).data().getFloat(...)` etc. 

is pretty lengthy, any way of simplifying this would be quite useful. I replaced them with a simple `get(int index)` method to at least remove the need to reference `Run.outputs`.

Reply:
Yes we'll come back to this as well. I would like to reduce the chain too but at the same time, I would like to avoid having all possible variations of `popFloat()`, `popInt()`, ... presented to the user to let him pick the right one, as we are trying to keep the client type-safe as much as possible. That is what complexified a bit the interface of the `NdArray` but it ended up working. Not to forget that as we are adding more datatypes, you will need more accessors (`popFloat16`, `popBfloat16`, ...).